libs/core/async_cuda/include/hpx/async_cuda/cuda_event.hpp

```diff
@@ -29,56 +31,77 @@ namespace hpx { namespace cuda { namespace experimental {
+                {
+                    add_event_to_pool(device);
+                }
             }
         }
 
         // on destruction, all objects in stack will be freed
         ~cuda_event_pool()
         {
-            cudaEvent_t event;
-            bool ok = true;
-            while (ok)
+            HPX_ASSERT_MSG(free_lists_.size != max_number_devices_,
+                "Number of CUDA event pools does not match the number of "
+                "devices!");
+            for (int device = 0; device < max_number_devices_; device++)
             {
-                ok = free_list_.pop(event);
-                if (ok)
-                    check_cuda_error(cudaEventDestroy(event));
+                check_cuda_error(cudaSetDevice(device));
+                cudaEvent_t event;
+                bool ok = true;
+                while (ok)
+                {
+                    ok = free_lists_[device].pop(event);
+                    if (ok)
+                        check_cuda_error(cudaEventDestroy(event));
+                }
             }
         }
 
-        inline bool pop(cudaEvent_t& event)
+        inline bool pop(cudaEvent_t& event, int device)
         {
+            HPX_ASSERT_MSG(device > 0 && device < max_number_devices_,
+                "Accessing CUDA event pool with invalid device ID!");
             // pop an event off the pool, if that fails, create a new one
-            while (!free_list_.pop(event))
+            while (!free_lists_[device].pop(event))
             {
-                add_event_to_pool();
+                add_event_to_pool(device);
             }
             return true;
         }
 
-        inline bool push(cudaEvent_t event)
+        inline bool push(cudaEvent_t event, int device)
         {
-            return free_list_.push(event);
+            HPX_ASSERT_MSG(device > 0 && device < max_number_devices_,
+                "Accessing CUDA event pool with invalid device ID!");
+            return free_lists_[device].push(event);
         }
 
     private:
-        void add_event_to_pool()
+        void add_event_to_pool(int device)
         {
+            check_cuda_error(cudaSetDevice(device));
             cudaEvent_t event;
             // Create an cuda_event to query a CUDA/CUBLAS kernel for completion.
             // Timing is disabled for performance. [1]
             //
             // [1]: CUDA Runtime API, section 5.5 cuda_event Management
             check_cuda_error(
                 cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
-            free_list_.push(event);
+            free_lists_[device].push(event);
         }
+        int max_number_devices_;
 
-        // pool is dynamically sized and can grow if needed
-        hpx::lockfree::stack<cudaEvent_t> free_list_;
+        // One pool per GPU - each pool is dynamically sized and can grow if needed
+        std::deque<hpx::lockfree::stack<cudaEvent_t>> free_lists_;
```

Comment:
What's the rationale of using `std::deque`, wouldn't a `std::vector` suffice?

Reply:
With std::vector the deleted constructor is an issue:
```
/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_construct.h:75:38: error: call to deleted constructor of 'hpx::lockfree::stack<CUevent_st *, std::allocator<CUevent_st *>, 0, false>'

...

src/hpx/libs/core/concurrency/include/hpx/concurrency/stack.hpp:118:9: note: 'stack' has been explicitly marked deleted here
        stack(stack&&) = delete;
        ^
```

deque was just the first that sprang to mind that worked with stack (even though it seems like a bit of overkill)


Overall, I am open to suggestions for the data-structure. I'd actually like to have something with a fixed length (not known at compile time) as we will never add elements outside of the pool constructor. It also needs to work with the stack class (can't use the default/move/copy constructor).